docs/Recovery-Postcard.md

```diff
@@ -0,0 +1,143 @@
+- Multiple PUK codes with its indexes, printed in order that user should use them.
+- A brief information about how to use the postcard.
+
+For example:
+![Recovery Postcard Example](resources/images/recovery-postcard-example.png)
+
+
+## Recovery Postcard Construction
+
+The purpose of this section is to describe how the postcard is securely exchanged between PowerAuth Server and the recovery postcard printing service. The following abbreviations will be used in this section:
+
+- **"PAS"** – stands for PowerAuth Server
+- **"RPPS"** – stands for "Recovery Postcard Printing Service". It's a service responsible for the postcard manufacturing.
+- **"RPOS"** – stands for "Recovery Postcard Ordering Service". It's a service that orchestrates a whole process, from postcard issue, to send order to RPPS
+
+### Introduction to the problem
+
+1. The PowerAuth Server is typically located in a controlled and very secure environment in the bank
+2. On opposite to that, the Printing Service responsible for Recovery Postcard manufacturing, is typically located outside of the bank.
+   - It's expected that the service has also some level of physical security (e.g. 24/7 security cameras), operator in factory has no physical access to the printer or the computer.
+3. We don't want to transmit recovery codes in plaintext (or even encrypted with the pre-shared key)
+
+
+### Printing order
+
+1. PAS and RPPS must have pre-shared asymmetric EC keys (P256r1), leading to 32bytes long shared secret (ECDH). Let's call this as `RPPS_MASTER_SHARED_SECRET`.
+
+1. RPOS asks PAS for postcard creation for one particular user
+
+1. PAS will calculate
+   ```java
+   byte[] SHARED_SECRET
+   if (RPPS_MASTER_SHARED_SECRET == null) {
+       // Mode useful for RC associated with activation
+       SHARED_SECRET = Generator.randomBytes(32)
+   } else {
+       // Mode useful for RC postcard creation
+       SHARED_SECRET = RPPS_MASTER_SHARED_SECRET
+   }
+   
+   repeat 
+   {
+       byte[] NONCE = Generator.randomBytes(32)
+       byte[] SHARED_POSTCARD = KDF_X9_63_SHA256.derive(SHARED_SECRET, NONCE)
+       byte[] RECOVERY_CODE_BYTES = ByteUtils.subarray(SHARED_POSTCARD, 0, 10)
+       String RECOVERY_CODE = Generator.buildActivationCode(RECOVERY_CODE_BYTES)
+       if RECOVERY_CODE doesn't exist in DB {
+           break
+       }
+   }
+
+   byte[] RECOVERY_PUK_BASE_KEY = ByteUtils.subarray(SHARED_POSTCARD, 10, 16)
+
+   // Prepare array of PUK derivation indexes, set to 0
+   long[] PUK_DERIVATION_INDEX_ARRAY[PUK_COUNT] = { 0 }
+   
+   // Prepare array of PUK values
+   String[] PUK_ARRAY[PUK_COUNT] = { null }
+
+   for I = 0; I < PUK_COUNT; i++ 
+   {
+       repeat
+       {
+           long PUK_DERIVATION_INDEX = ByteUtils.getLong(Generator.randomBytes(8))
+           byte[] PUK_KEY = KDF.derive(RECOVERY_PUK_BASE_KEY, PUK_DERIVATION_INDEX)
+           long PUK_VALUE = (ByteUtils.getLong(ByteUtils.truncate(PUK_KEY, 8)) & 0xFFFFFFFFFF) % Math.pow(10,10)
```

Comment:
Is the `0xFFFFFFFFFF` constant correct, or should there be `0x7FFFFFFFFF` (striping the signature bit)?

Reply:
I believe it is correct, the `long` is 8-byte (at least in Java), 0xFFFFFFFFFF takes 5 lowest bytes. So the first 3 bytes from the long value are thrown away including the signature bit.

Note that in the actual implementation we use:
```
        // Extract last 8 bytes from PUK key bytes
        byte[] truncatedBytes = new byte[8];
        System.arraycopy(pukKeyBytes, 8, truncatedBytes, 0, 8);
        long puk = (ByteBuffer.wrap(truncatedBytes).getLong() & 0xFFFFFFFFFFL) % (long) (Math.pow(10, 10));
```

Notice the `getLong()` method and `L` in `0xFFFFFFFFFFL`, this is necessary because we work with 5 bytes and not with 4 like in case of other similar operations in PowerAuth crypto.

The usual masking in PowerAuth is 4-byte and based on `getInt()`,  e.g.:
```
int index = signatureLong.length - 4;
int number = (ByteBuffer.wrap(signatureLong).getInt(index) & 0x7FFFFFFF) % (int) (Math.pow(10, PowerAuthConfiguration.SIGNATURE_LENGTH));
```

Reply:
The value is 64 bit, so constant `0x7FFFFFFFFF` is not what you think it is :)  I've selected a constant which is most close to 10^10, but byte-aligned - `0xFF FF FF FF FF`. I'll put a `L` postfix to the pseudocode, to highlight that it's a long value.